include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,356 @@
+                                __out_values[__i] = __binary_op(__carry_in, __out_values[__i]);
+
+                                if (__i == __n - 1 || __keys[__i] != __keys[__i + 1])
+                                    break;
+                            }
+                        }
+                    }
+                    else // exclusive scan
+                    {
+                        if (__local_id != 0)
+                        {
+                            for (int32_t __i = __start; __i < __end; ++__i)
+                            {
+                                __out_values[__i] = __binary_op(__carry_in, __out_values[__i]);
+
+                                if (__i == __n - 1 || __keys[__i] != __keys[__i + 1])
+                                    break;
+                            }
+                        }
+                    }
+
+                    if (__local_id == __wgroup_size - 1) // last work item writes the group's carry out
+                    {
+                        auto __group_id = __group.get_group_id(0);
+
+                        // If no segment ends in the item, the aggregates from previous work groups must be applied.
+                        if (__max_end == 0)
+                            __partials_acc[__group_id] = __binary_op(__carry_in, __accumulator);
+
+                        else
+                            __partials_acc[__group_id] = __accumulator;
+
+                        __flag_type __seg_present = __closest_seg_id != 0;
+                        __seg_ends_acc[__group_id] = __seg_present;
+                    }
+                });
+        });
+
+        // 2. Apply work group carry outs, calculate output indices, and load results into correct indices.
+        __exec.queue()
+            .submit([&](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_values);
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_scan);
+
+                __cgh.parallel_for(
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        auto __group_id = __group.get_group_id(0);
+                        ::std::size_t __global_id = __item.get_global_id();
+                        auto __local_id = __item.get_local_id();
+                        auto __start = __global_id * __vals_per_item;
+                        auto __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                        int32_t __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector{};
+                        
+                        // 2a. Calculate the work group's carry-in value.
```

Comment:
This section (2a) uses thread 0 by itself to calculate the accumulated carry in value of all previous workgroups involved in the current segment.  For most segment sizes this is minimal work, which spans a small number of previous workgroups.

However, for workgroups beginning at the end of a very large segment that spans many workgroups, this can be a non-trivial amount of work that is done by a single thread, with the rest of the workgroup waiting.

The trouble is, we don't know how large the segment is, or how many previous workgroups it spans until we are done.  We need to decide if any special detection / handling of very large segments is necessary, it depends on our target use case.  

Its likely that any attempt to fix this will drag down the "normal" case.

Reply:
This was with a commit to make this section parallel (there still exists a small serial section, but it is acceptable).